File: 6.ZigZagConversion/Solution.java
import java.util.ArrayList;
import java.util.List;

public class Solution {
    // Use string builder instead
    public String convert(String s, int numRows) {
        if (s.length() < numRows || numRows == 1 || s.length() < 2)
            return s;
        int step = 1, index = 0;
        List<String> res = new ArrayList<String>(numRows);
        String ans = "";
        for (int i = 0; i < s.length(); i++) {
            res.set(index, res.get(index) + s.charAt(i));
            if (index == 0)
                step = 1;
            else if (index == numRows - 1)
                step = -1;
            if (step == 1)
                index++;
            else
                index--;
        }
        for (String x : res) {
            ans += res;
        }
        return ans;
    }
}

File: 6.ZigZagConversion/Solution.cpp
#include <bits/stdc++.h>
using namespace std;
class Solution
{
public:
    string covert(string s, int n)
    {
        if (s.length() < 2 || n >= s.length())
            return s;
        vector<string> res(n, "");
        int step = 1, index = 0;
        for (int i = 0; i < s.length(); i++)
        {
            res[index].push_back(s[i]);
            if (index == 0)
                step = 1;
            else if (index == n - 1)
                step = -1;
            if (step == -1)
                index--;
            else
                index++;
        }
        string ans = "";
        for(string x : res){
            ans += x;
        }
        return ans;
    }
};